src/model/openthaigpt_pretraining_model/models/llama_hf/model.py

```diff
@@ -0,0 +1,508 @@
+
+        loss = None
+        if labels is not None:
+            # Shift so that tokens < n predict n
+            shift_logits = logits[..., :-1, :].contiguous()
+            shift_labels = labels[..., 1:].contiguous()
+            # Flatten the tokens
+            loss_fct = CrossEntropyLoss()
+            shift_logits = shift_logits.view(-1, self.config.vocab_size)
+            shift_labels = shift_labels.view(-1)
+            # Enable model parallelism
+            shift_labels = shift_labels.to(shift_logits.device)
+            loss = loss_fct(shift_logits, shift_labels)
+
+        if not return_dict:
+            output = (logits,) + outputs[1:]
+            return (loss,) + output if loss is not None else output
+
+        return CausalLMOutputWithPast(
+            loss=loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
+
+    def prepare_inputs_for_generation(
+        self,
+        input_ids,
+        past_key_values=None,
+        attention_mask=None,
+        inputs_embeds=None,
+        **kwargs,
+    ):
+        if past_key_values:
+            input_ids = input_ids[:, -1:]
+
+        position_ids = kwargs.get("position_ids", None)
+        if attention_mask is not None and position_ids is None:
+            # create position_ids on the fly for batch generation
+            position_ids = attention_mask.long().cumsum(-1) - 1
+            position_ids.masked_fill_(attention_mask == 0, 1)
+            if past_key_values:
+                position_ids = position_ids[:, -1].unsqueeze(-1)
+
+        # if `inputs_embeds` are passed, we only want to use them in the 1st generation step # noqa
+        if inputs_embeds is not None and past_key_values is None:
+            model_inputs = {"inputs_embeds": inputs_embeds}
+        else:
+            model_inputs = {"input_ids": input_ids}
+
+        model_inputs.update(
+            {
+                "position_ids": position_ids,
+                "past_key_values": past_key_values,
+                "use_cache": kwargs.get("use_cache"),
+                "attention_mask": attention_mask,
+            }
+        )
+        return model_inputs
+
+    @staticmethod
+    def _reorder_cache(past_key_values, beam_idx):
+        reordered_past = ()
+        for layer_past in past_key_values:
+            reordered_past += (
+                tuple(
+                    past_state.index_select(0, beam_idx) for past_state in layer_past
+                ),
+            )
+        return reordered_past
+
+
+def make_model_llama(vocab_size, context_length, use_checkpointing):
+    cfg = LlamaConfig(
+        vocab_size=vocab_size,
+        hidden_size=1024,
+        num_hidden_layers=8,
+        num_attention_heads=8,
+        hidden_act="silu",
+        max_position_embeddings=context_length,
+    )
+    if use_checkpointing == 2:
```

Comment:
I think you should feed `use_checkpointing` and `checkpoint_only_attention` in the input of model. You use `use_checkpointing==2` we don't know how different between `0`, `1`, and `2`.

Reply:
I change It in utils.py now to use enum instead, Must I do this?

Reply:
For me I will write
```python
if use_checkpointing and checkpoint_only_attention:
    model = LlamaForCausalLMNewCheckpoint(cfg)
    print("use gradient checkpointing only attentions")
else:
    model = LlamaForCausalLM(cfg)
```

Is it clearer than `use_checkpointing=2`?
You can pass 2 parameters without using Enum and some login in above.

Reply:
good Idea, done.